How to suspend - a guide

scarletfog opened the pull request:
### Context
There is a new feature that can be used - suspend/bach. Present it and give some examples/code snippers how to use it and what are the differences between API methods.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Improvement (better description, more examples etc.)

### Related issue(s):
1. https://github.com/handsontable/docs/issues/261

scarletfog commented:
@wojciechczerniak I need some feedback on this one 🙏

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+            }
+          </style>
+          <script data-jsfiddle="example2">
+            let now = null;
+            function startLog() {
```
```suggestion
            let now = null;

            function startLog() {
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+
+            var data2 = [
+              [11, 'Gavin Elle', 'Floppy socks', 'XS', 3, 'yes'],
+            ];
+            var data3 = [
```
```suggestion
            ];

            var data3 = [
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+              colHeaders: ['ID', 'Customer name', 'Product name', 'Size', 'qty', 'Return'],
+            });
+
+            function alterTable() {
+            startLog();
+              hot.alter('insert_row', 10, 10);
```
```suggestion
            function alterTable() {
              startLog();
              hot.alter('insert_row', 10, 10);
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+              hot.setCellMeta(10, 2, 'className', 'green-bg');
+              hot.alter('remove_col', 6, 1);
+              hot.setCellMeta(0, 5, 'className', 'red-bg');
+              hot.setCellMeta(10, 5, 'className', 'red-bg');
+            printLog();
+            }
+            alterTable();
```
```suggestion
              hot.setCellMeta(10, 5, 'className', 'red-bg');
              printLog();
            }

            alterTable();
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+          </script>
+      </div>
+
+      <p>No<code>batch</code> used. The results are very same but it takes longer.
```
```suggestion
      <p>No <code>batch</code> used. The results are very same but it takes longer.
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+            });
+
+            function alterTable() {
+            startLog();
+            hot.batch(() => {
+              hot.alter('insert_row', 10, 10);
+              hot.alter('insert_col', 6, 1);
+              hot.populateFromArray(10, 0, data2);
+              hot.populateFromArray(11, 0, data3);
+              hot.setCellMeta(2, 2, 'className', 'green-bg');
+              hot.setCellMeta(4, 2, 'className', 'green-bg');
+              hot.setCellMeta(5, 2, 'className', 'green-bg');
+              hot.setCellMeta(6, 2, 'className', 'green-bg');
+              hot.setCellMeta(8, 2, 'className', 'green-bg');
+              hot.setCellMeta(9, 2, 'className', 'green-bg');
+              hot.setCellMeta(10, 2, 'className', 'green-bg');
+              hot.alter('remove_col', 6, 1);
+              hot.setCellMeta(0, 5, 'className', 'red-bg');
+              hot.setCellMeta(10, 5, 'className', 'red-bg');
+            })
+            printLog();
```
Code indentation?

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+            })
+            printLog();
+            }
+            // alterTable();
```
Remove this line if not needed? But looks like should be enabled?

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+
+            var data2 = [
+              [11, 'Gavin Elle', 'Floppy socks', 'XS', 3, 'yes'],
+            ];
+            var data3 = [
```
```suggestion
            ];

            var data3 = [
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+            }
+        </style>
+        <script data-jsfiddle="example1">
+            let now = null;
+            function startLog() {
```
```suggestion
            let now = null;

            function startLog() {
```

wojciechczerniak commented on tutorials/suspending-operations.html:
```diff
@@ -0,0 +1,348 @@
+      <pre>
+        <code class="lang-js">
+hot.batchRender(() => {
+hot.alter('insert_row', 5, 45);
+hot.setDataAtCell(1, 1, 'x');
+// The table will be rendered once after executing the callback
```
Content of the closure should have indentation?
```suggestion
  hot.alter('insert_row', 5, 45);
  hot.setDataAtCell(1, 1, 'x');
  // The table will be rendered once after executing the callback
```

scarletfog replied:
I think it will stand out slightly, for instance, in the migration guide we have:
![image](https://user-images.githubusercontent.com/20072336/103340184-c1f28600-4a83-11eb-890f-09b439e02406.png)

and if I make an indent here:

![image](https://user-images.githubusercontent.com/20072336/103340245-e0f11800-4a83-11eb-996a-fc9ea64bf8d2.png)

wojciechczerniak replied:
But you have `batchRender` here with a closure, so it will look more like:

```
hot.batchRender(() => {
    hot.alter('insert_row', 5, 45);
    hot.setDataAtCell(1, 1, 'x');
```

scarletfog replied:
Ah yes, maybe it's good to read from the top to bottom @scarletfog!

I will correct that...

wojciechczerniak reviewed (changes requested):
Part 1. First looks review. 
Part 2. I will now run and read the doc.

scarletfog commented:
@wojciechczerniak can you please have a look again?

wojciechczerniak commented:
So, I've read the page top to bottom. Content is great, API explained perfectly. But that is also the main issue I see. From the first section we're focusing on methods, not outcomes. While at the beginning we should explain "why's".

Menu and page name "Suspending operations" is misleading, we're not suspending CRUD operations. And the first section is problematic, as noted above, we're explaining APIs. While, we should explain that CRUD operations are triggering `render()` and if our clients combine multiple of them in series there will be unnecessary `render()` calls, while only the last one reveals the final output. The idea is to reduce those, and have a performance gain. We can use the same visualization chart that was created for blog post. Move the API introdution to the top of "API methods" section.

"API methods" section is ok, with the remarks below.
"Live demo of suspend feature" looks good. I would combine the two, there is a lot duplicated code. I think that we need two `alterTable` methods / buttons. One with `batch` and other without. Or buttons with `onclick` events, one will call pure `alterTable()` the other will wrap it in `batch(() => { alterTable() })`.

---

When it comes to the APIs I would switch the order. `batch` should be first and recommended in most cases. Then `batch*` methods if someone needs more control, and if batched operations include async methods then `suspend*`/`resume*`(example does not have async to support the idea/problem with `batch*`).

scarletfog commented:
@wojciechczerniak I haven't addressed all of the suggestions yet (one demo + async example) but could you please check if the content goes in a good direction?

I changed the sub-sections in the API section - now it is divided into batch* vs suspend* but I am not 100% sure it better.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+    <p>
+      There are several <a href="#page-suspend-api">API methods</a> you can use for suspending but the <code>batch</code> is the most universal one.
+      It is a callback function - the render is executed after all operations provided inside of the body are completed.
+      This is a recommended method as least prone to human errors.
```
To be positive, we would say that this is the recommended method because it will take care of everything for the user and is the safest and easiest to use. Just place your operations inside a closure, Handsontable will take care of the suspending and render a single time in the end.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+      <br>
+    </p>
+    <p>
+      Note that this is an advanced feature which can improve the performance but should be used with caution.
+      Misuse can result in some flickering, glitches, or other visual distortion.
```
`batch` should be safe, we're recommending it just a paragraph before. Maybe we should say that other API methods are more advanced and should be used with caution? This is true when someone calls `suspend` few times and forgets the `resume` or will mix `*render` with `*operations*`.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+      <li>both rendering and the execution.</li>
+    </ul>
+
+    <br>
+      Term "rendering" refers directly to DOM rendering, the term
+      "execution" refers to all operations that are different from DOM rendering.
+      Currently only the indexes recalculation allows to postpone the process.
+    </p>
+
+    <p>
+      Methods which names start with <strong>batch*</strong> prefix, that is: <code>batch</code>,<code>batchRender</code> and <code>batchExecution</code>
+      are recommended to be used as the first choice if <strong>you don't need to batch async operations</strong>.
+      <br/>
+      Methods which names start with <strong>suspend*</strong> prefix, that is: <code>suspendRender</code> and <code>suspendExecution</code>
+      are of the second choice - useful when you need to batch async operations. In essence they work the same as <strong>batch*</strong> methods
+      but the <strong>render has to be resumed manually</strong>.
+    </p>
```
![Screenshot 2021-01-08 at 17 13 19](https://user-images.githubusercontent.com/4718998/104037348-e4e40f00-51d4-11eb-9e90-a3c2b9f73bd2.png)

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+    </p>
+    <p>
+      <p><strong>batchRender</strong> </p>
+      The <code>batchRender</code> method is less prone to human errors since the operations will be calculated after executing the callback.
```
`batchRender` says about resuming operations. Wrong copy&paste?

Also, the first sentence should explain what it does. Not that it's less prone to errors.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+  </p>
+  <p>
+    <p><strong>batchExecution</strong> </p>
+    The <code>batchExecution</code> method is less prone to human errors since the operations will be calculated after executing the callback.
```
Would be nice if the first sentence would explain what it does.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,401 @@
+      It is a callback function - the render is executed after all operations provided inside of the body are completed.
+      This is a recommended method as least prone to human errors.
+      <br/>
+      The following snippet shows a simple example of a few operations batched:
```
Explain: we call three API operations, each calls `render()` afterwards. By placing them inside a `batch` callback we will skip two renders, which is super optimal.

wojciechczerniak reviewed (changes requested):
There is something wrong with examples. I had to comment them out to run the page. They go in some kind of a loop?

scarletfog commented:
@wojciechczerniak it is buggy only for the example with the `batch` method. If I remember well when I was working with the hot version based on budzio's branch (with the suspend feature) it worked OK - I will double-check to be sure. At some point I had the `alterTable();` function commented just to keep focused on the content.

I will address the suggestions first.

scarletfog commented:
@wojciechczerniak please have a look. I will work on improving the example in the meantime.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,410 @@
+      Note that there are other methods that can be used to batch operations, however they are slightly more advanced and should
+      be used with caution. Flickering, glitches, or other visual distortion may happen when you forget to <code>resume</code> render after
+      suspending it several times. Mixing methods of a render type with those focused on operations can also result in some unexpected behavior.
+
```
Should we add the last sentence to strengthen the message: "`batch` should be sufficient in most use cases"?

wojciechczerniak reviewed (changes requested):
Thanks for addressing all my comments. We should fix the examples and that's it

budnix commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,408 @@
+      The diagram shows a comparison, the same operations were performed with (deep blue columns) and without the batch (light blue columns).
+      The gain in speed of execution time increases with the number of operations batched.
+    </p>
+      <img src="images/batch_operations_comparison.png" alt="batch_diagram" />
```
I've built the docs and I've seen that the image with charts isn't perfectly fitted to the page. As a result, it's blurred. Can I suggest adding an image exactly in size or two times bigger than the placeholder size? So, it'd be 1434x990px.

scarletfog replied:
Thank you so much for your help @budnix ! So much appreciated 😄 

About the image - I will ask Mariusz.

budnix reviewed (commented):
I've read the article, the generated version on my local machine. It's understandable and very enjoyable to read 👏 

I allowed myself to commit my suggestion directly to the branch as there were some issues related to multiple places. In the commit, I've fixed the second demo which won't work, and rewrite the code to ES5 so that it should work on IE.

scarletfog commented:
@budnix could you please check if the image looks better? :) 

@wojciechczerniak one question since I don't remember what was the decision - should we keep "Suspend rendering" as the title inside the menu or change it to "Batching operations"?

budnix commented:
> @budnix could you please check if the image looks better? :)

It's better 👍 
![Chart-comp](https://user-images.githubusercontent.com/571316/105345276-5a43dc80-5be4-11eb-9d82-1ff332568673.png)

wojciechczerniak commented:
> @wojciechczerniak one question since I don't remember what was the decision - should we keep "Suspend rendering" as the title inside the menu or change it to "Batching operations"?

Suspend rendering. "render" is a keyword here when someone has to optimize their code.

scarletfog commented:
@wojciechczerniak one last question! I remember you wanted to make these two examples into one. Do you have something specific in mind?

wojciechczerniak commented:
> @wojciechczerniak one last question! I remember you wanted to make these two examples into one. Do you have something specific in mind?

Yeah, single Handsontable instance. Two buttons. With logging the time

```tsx
function alterTable() {
 // all ops
}

function standardAlterTable() {
  const t1 = performance.now();
  alterTable();
  const t2 = perofrmance.now();

  logOutput('[without batch] ${t2 - t1} ms');
}

function batchedAlterTable() {
  const t1 = performance.now();
  hot.batch(alterTable);
  const t2 = perofrmance.now();

  logOutput('[with batch] ${t2 - t1} ms');
}

function logOutput(msg) {
  const logDiv = document.querySelector('#logOuput');
  const div = document.createElement('div');
  div.innerText = msg;
  logDiv.append(div);
}

<div id="example"></div>
<button onclick="standardAlterTable">without batch</button>
<button onclick="batchedAlterTable">with batch</button>
<div id="logOutput"></div>
```

budnix commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,363 @@
+              alterTable();
+              const t2 = performance.now();
+
+              logOutput(`Time without batch ${t2 - t1} ms`);
```
```suggestion
              logOutput(`Time without batch ${(t2 - t1).toFixed(2)} ms`);
```
With a fixed number, it looks a little bit better. What do you think?

budnix replied:
As the code has to be rewritten to ES3 I make a change within https://github.com/handsontable/docs/pull/267/commits/534c5de25c9f65c4afe4b8f1f5153e2dfab40273.

budnix commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,363 @@
+              hot.batch(alterTable);
+              const t2 = performance.now();
+
+              logOutput(`Time with batch ${t2 - t1} ms`);
```
```suggestion
              logOutput(`Time without batch ${(t2 - t1).toFixed(2)} ms`);
```

budnix replied:
Changes within https://github.com/handsontable/docs/pull/267/commits/534c5de25c9f65c4afe4b8f1f5153e2dfab40273

budnix commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,363 @@
+              logOutput(`Time with batch ${t2 - t1} ms`);
+            }
+
+            function logOutput(msg) {
+              const logDiv = document.querySelector('#logOuput');
+              const div = document.createElement('div');
+              const now = new Date();
+
+              div.innerText = `[${now.toTimeString().slice(0,8)}] ${msg}`;
+              logDiv.prepend(div);
+            }
```
```suggestion
            function logOutput(msg) {
              var logDiv = document.querySelector('#logOuput');
              var div = document.createElement('div');
              var now = new Date();

              div.innerText = '[' + now.toTimeString().slice(0,8) + ']' ' + msg;
              logDiv.prepend(div);
            }
```
ES3 style :)

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -225,24 +228,45 @@ <h3 id="page-suspend-demo">Live demo of suspend feature</h3>
-              now = null;
+            #logOuput {
+              padding: 15px 0 10px 0;
+              overflow-y: scroll;
```
"auto", no point in showing scroll until we will reach max height?

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -266,137 +290,69 @@ <h3 id="page-suspend-demo">Live demo of suspend feature</h3>
               data: data1,
               width: 700,
-              height: 300,
+              height: 250,
```
That is not enough to show the whole table
We have a limited number of rows, no point in limiting the height

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -266,137 +290,69 @@ <h3 id="page-suspend-demo">Live demo of suspend feature</h3>
-              background-color: #2ECC40;
-            }
+            Handsontable.dom.addEvent(buttonWith, 'click', function() {
+              batchedAlterTable();
```
Maybe we could move the function body here? Might be clearer to read, and shorter as an example

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -266,137 +290,69 @@ <h3 id="page-suspend-demo">Live demo of suspend feature</h3>
+            function logOutput(msg) {
+              const logDiv = document.querySelector('#logOuput');
+              const div = document.createElement('div');
+              const now = new Date();
```
I think that @budnix few times was fixing examples to use `var` for backward compatibility.

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -314,6 +314,7 @@ <h3 id="page-suspend-demo">Live demo of suspend feature</h3>
               hot.alter('remove_row', 10, 10);
               hot.setCellMeta(0, 5, 'className', 'red-bg');
               hot.setCellMeta(10, 5, 'className', 'red-bg');
+              hot.render(); // Render here is needed to populate the new "className"s
```
I think that `standardAlterTable` will be a better place for render.
In `batchAlterTable` it is redundant since `batch` will call the render at the end.

budnix replied:
It's irrelevant for "batch" mode. Here, we show that code that works without batching can be easily copy-pasted to a batch callback. In this way, no one will tell us that in the "batchedAlterTable" we forgot (or even lie) that there is missing a "render" call.

scarletfog replied:
I have already moved it 😅

wojciechczerniak replied:
Right, `batch` will suspend it anyway. Good suggestion with the copy&paste to `batch` callback. 
"It's the same code, just wrap it and eat the fruits from the tree of performance!"

budnix replied:
Thanks! I restored the "render" call within https://github.com/handsontable/docs/pull/267/commits/273cdaadfe51267ab0718f8140440b69717d140a.

budnix commented:
I don't why, but the table within Suspend Render looks different than the table from other pages:
![image](https://user-images.githubusercontent.com/571316/105851830-e3dc1b80-5fe3-11eb-940c-9c7986c88d23.png)

wojciechczerniak commented on tutorials/suspend-rendering.html:
```diff
@@ -0,0 +1,351 @@
+
+            #logOuput {
+              padding: 15px 0 10px 0;
+              max-height:400px;
```
```suggestion
              max-height:400px;
              overflow-y: auto;
```

<img width="660" alt="Screenshot 2021-01-26 at 14 45 51" src="https://user-images.githubusercontent.com/4718998/105852848-4255c980-5fe5-11eb-863c-f5388d28168b.png">

budnix commented:
> I don't why, but the table within Suspend Render looks different than the table from other pages:
> ![image](https://user-images.githubusercontent.com/571316/105851830-e3dc1b80-5fe3-11eb-940c-9c7986c88d23.png)

I've found that there is a missing `hot` class. Fixed https://github.com/handsontable/docs/pull/267/commits/273cdaadfe51267ab0718f8140440b69717d140a.

scarletfog commented:
Thanks @budnix !!!

wojciechczerniak reviewed (approved):
🏆

scarletfog commented:
@budnix anything missing here or should I give it a MERGE? 😄

budnix reviewed (approved):
👏
